What's New in 2.0 (#4165)
sunpy/sunpy: docs/whatsnew/2.0.rst

.. doctest-skip-all

.. _whatsnew-2.0:

************************
What's New in SunPy 2.0?
************************

Overview
========

The SunPy project is pleased to announce the 2.0 release of the sunpy package.

> .. _whatsnew-2.0-pixels:

Reviewer: wait what's this doing here.

In 2.0 the way you call both methods has changed, to accept a bottom left and then either width and height or a top right coordinate.
As part of this change, the ``top_right``, ``width``, and ``height`` arguments must always be keyword arguments, i.e. ``width=10*u.arcsec``.

This change allows you to give the same rectangle specification to `~sunpy.map.GenericMap.submap` as to `~sunpy.map.GenericMap.draw_rectangle`.
Which is especially useful when you wish to plot a cropped area of a map, along with it's context in the parent map::

    >>> import astropy.units as u
    >>> from astropy.coordinates import SkyCoord
    >>> import matplotlib.pyplot as plt

    >>> import sunpy.map
    >>> from sunpy.data.sample import AIA_171_IMAGE

    >>> aia = sunpy.map.Map(AIA_171_IMAGE)

    >>> bottom_left = SkyCoord(-100 * u.arcsec, -100 * u.arcsec, frame=aia.coordinate_frame)
    >>> width = 500 * u.arcsec
    >>> height = 300 * u.arcsec

    >>> sub_aia = aia.submap(bottom_left, width=width, height=height)

    >>> fig = plt.figure()

>     >>> ax1 =fig.add_subplot(1, 2, 1, projection=aia)

Reviewer: ```suggestion
    >>> ax1 = fig.add_subplot(1, 2, 1, projection=aia)
```

Our `~sunpy.coordinates.frames.HeliographicCarrington` coordinate frame is now expressly intended for the co-alignment of images of the Sun's surface from different observatories.
`~sunpy.coordinates.frames.HeliographicCarrington` now requires the specification of the observer location (Earth or otherwise) because the light travel time between the Sun and the observer is accounted for.
SunPy output now matches the calculations by `JPL Horizons <https://ssd.jpl.nasa.gov/?horizons>`__ and `SPICE <https://naif.jpl.nasa.gov/naif/>`__.
There may be small differences compared to Carrington coordinates computed by groups that do not use modern parameter values or the same assumptions for the methodology.

Importantly, the Carrington longitude that is now calculated (including using :func:`sunpy.coordinates.sun.L0`) will not match earlier versions of SunPy.
A detailed write-up of the calculation approach and comparisons to other resources can be found :ref:`here <sunpy-coordinates-carrington>`.

Download behind proxies
=======================

> With the release of parfive 1.1, sunpy has been patched now to able to pass proxy information to `aiohttp.ClientSession`.

Reviewer: ```suggestion
With the release of parfive 1.1, sunpy has been patched to be able to utilize proxy servers when downloading files.
```

* Proxy URL is read from the environment variables `HTTP_PROXY` or `HTTPS_PROXY`.
* Proxy Authentication `proxy_auth` should be passed as a `aiohttp.BasicAuth` object, explicitly by the user.
* Proxy Headers `proxy_headers` should be passed as `dict` object, explicitly by the user.

For example if you use a bash terminal:

.. code-block:: bash

    $ HTTP_PROXY=http://user:password@proxyserver.com:3128
    $ HTTPS_PROXY=https://user:password@proxyserver.com:3128
    $ export HTTP_PROXY
    $ export HTTPS_PROXY

these will be used to enable downloads through a proxy.

Citation update
===============

A paper discussing sunpy 1.0 was accepted in The Astrophysical Journal and you can find the bibtex for it by running::

    >>> import sunpy
    >>> sunpy.__citation__

or `accessing the website directly <https://iopscience.iop.org/article/10.3847/1538-4357/ab4f7a>`__.

> aiaprep is now deprecated

Reviewer: now I think this is too low down the page :see_no_evil:

Reviewer 2: Yeah I'd agree. It is a small change, but something that people obviously will want to know about

Changes to Carrington coordinates
=================================

We have refined our approach for heliographic Carrington coordinates to best support high-resolution imagery of the Sun, including from observatories that are at distances from the Sun that is significantly different from 1 AU (e.g., `Solar Orbiter <https://en.wikipedia.org/wiki/Solar_Orbiter>`__).
Our `~sunpy.coordinates.frames.HeliographicCarrington` coordinate frame is now expressly intended for the co-alignment of images of the Sun's surface from different observatories.
`~sunpy.coordinates.frames.HeliographicCarrington` now requires the specification of the observer location (Earth or otherwise) because the light travel time between the Sun and the observer is accounted for.
SunPy output now matches the calculations by `JPL Horizons <https://ssd.jpl.nasa.gov/?horizons>`__ and `SPICE <https://naif.jpl.nasa.gov/naif/>`__.
There may be small differences compared to Carrington coordinates computed by groups that do not use modern parameter values or the same assumptions for the methodology.

> Importantly, the Carrington longitude that is now calculated (including using :func:`sunpy.coordinates.sun.L0`) with not match earlier versions of SunPy.

Reviewer: ```suggestion
Importantly, the Carrington longitude that is now calculated (including using :func:`sunpy.coordinates.sun.L0`) will not match earlier versions of SunPy.
```

    >>> bottom_left, top_right = get_rectangle_coordinates(bottom_left_vector)
    >>> bottom_left, top_right
    (<SkyCoord (HeliographicStonyhurst: obstime=None): (lon, lat, radius) in (deg, deg, km)
        (0., 0., 695700.)>,
    <SkyCoord (HeliographicStonyhurst: obstime=None): (lon, lat, radius) in (deg, deg, km)
        (0.00277778, 0.00277778, 695700.)>)

    # With bottom left as a BaseCoordinateFrame instance, width and height
    >>> bottom_left =  HeliographicStonyhurst(0 * u.arcsec, 0 * u.arcsec)
    >>> width = 10 * u.arcsec
    >>> height = 10 * u.arcsec
    >>> bottom_left, top_right = get_rectangle_coordinates(bottom_left, width=width, height=height)
    >>> bottom_left, top_right
    (<HeliographicStonyhurst Coordinate (obstime=None): (lon, lat, radius) in (deg, deg, km)
        (0., 0., 695700.)>,
    <HeliographicStonyhurst Coordinate (obstime=None): (lon, lat, radius) in (deg, deg, km)
        (0.00277778, 0.00277778, 695700.)>)

Download behind proxies
=======================

With the release of parfive 1.1, sunpy has been patched now to able to pass proxy information to `aiohttp.ClientSession`.

> * Proxy URL is read from the environment variables `HTTP_PROXY` or `HTTPS_PROXY`, depending on the protocol of the `url` passed.

Reviewer: ```suggestion
* Proxy URL is read from the environment variables `HTTP_PROXY` or `HTTPS_PROXY`.
```

* Proxy Authentication `proxy_auth` should be passed as a `aiohttp.BasicAuth` object, explicitly by the user.
* Proxy Headers `proxy_headers` should be passed as `dict` object, explicitly by the user.

For example if you use a bash terminal:

.. code-block:: bash

    $ HTTP_PROXY=http://user:password@proxyserver.com:3128
    $ HTTPS_PROXY=https://user:password@proxyserver.com:3128
    $ export HTTP_PROXY
    $ export HTTPS_PROXY

these will be used to enable downloads through a proxy.

Citation update
===============

> We had a new paper accepted in, The Astrophysical Journal and you can find the bibtex for it by::

Reviewer: ```suggestion
A paper discussing sunpy 1.0 was accepted in The Astrophysical Journal and you can find the bibtex for it by running::
```

    >>> import sunpy
    >>> sunpy.__citation__

or `accessing the website directly <https://iopscience.iop.org/article/10.3847/1538-4357/ab4f7a>`__.

aiaprep is now deprecated
=========================

With the release of the new `aiapy <https://aiapy.readthedocs.io>`__ package, `sunpy.instr.aia.aiaprep` will be removed in version 2.1.

> You can use the `register` function in the aiapy package to convert AIA images to level 1.5.

Reviewer: @wtbarnes is the wording here correct? I would have thought you needed to run the other functions in aiaprep to get close to 1.5? Is something like "to align AIA images" better?

Reviewer 2: ```suggestion
Equivalent functionality is provided by the `register` function in aiapy. For more details,
see the `example on registering and aligning level 1 AIA images <https://aiapy.readthedocs.io/en/latest/generated/gallery/prepping_level_1_data.html#registering-and-aligning-level-1-data>`_
in the aiapy documentation.
```

> Search Attributes

Reviewer: I propose we stick this at the top. It's probably our largest feature this release.

=================

To search with `~sunpy.net.Fido`, you need to specify attributes to search against.
Before sunpy 2.0, you had to supply values, as the following example demonstrates::

    >>> from sunpy.net import Fido, attrs as a
    >>> Fido.search(a.Time('2012/3/4', '2012/3/6'), a.Instrument("norh"), a.Wavelength(17*u.GHz))

There was no way to know if the value was correct, but now we have a extenstive list of supported values from the clients and servers we can request data from.

> Using ``Instrument`` as an example, if you print the object::

Reviewer: ```suggestion
Using `~sunpy.net.attrs.Instrument` as an example, if you print the object::
```

.. doctest-skip-all

.. _whatsnew-2.0:

************************
What's New in SunPy 2.0?
************************

Overview
========

The SunPy project is pleased to announce the 2.0 release of the sunpy package.

.. _whatsnew-2.0-pixels:

Fixes and clarification to pixel indexing
=========================================

> 

Reviewer 3: we should also explicitly call out `make_fitswcs_header`

Reviewer: Didn't we backport that fix? If we did it's not really a what's new in 2.0 as we already released it.

sunpy uses zero-based indexing when referring to pixels, where the center of
the bottom left pixel of a map is at ``[0, 0] * u.pix``. Several parts of the
API have been updated to make sure this is consistently the case across the
pacakge. In particular:

- `sunpy.map.GenericMap.top_right_coord` previously had an off-by-one error
  in the calculation of the top right coordinate. This has been fixed.
- `sunpy.map.GenericMap.center` previously had an off-by-one error
  in the calculation of the coordinate of the center of a map. This has been fixed.

> - `sunpy.map.GenericMap.reference_pixel` now returns a zero-based reference

Reviewer: are you also going to add notes about submap?

Author: I was thinking of, but didn't since I don't think there were any pixel indexing issues in submap.

Reviewer: Well we will probably have another section on rectangles, so we could mention it there maybe :smile:

Author: That sounds good 👍
